upb/util/required_fields.c

```diff
@@ -0,0 +1,294 @@
+ * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "upb/util/required_fields.h"
+
+#include <inttypes.h>
+#include <setjmp.h>
+#include <stdarg.h>
+#include <stdio.h>
+
+#include "upb/reflection.h"
+
+// Must be last.
+#include "upb/port_def.inc"
+
+////////////////////////////////////////////////////////////////////////////////
+// upb_FieldPath_ToText()
+////////////////////////////////////////////////////////////////////////////////
+
+typedef struct {
+  char* buf;
+  char* ptr;
+  char* end;
+  size_t overflow;
+} upb_PrintfAppender;
+
+UPB_PRINTF(2, 3)
+static void upb_FieldPath_Printf(upb_PrintfAppender* a, const char* fmt, ...) {
+  size_t n;
+  size_t have = a->end - a->ptr;
+  va_list args;
+
+  va_start(args, fmt);
+  n = vsnprintf(a->ptr, have, fmt, args);
+  va_end(args);
+
+  if (UPB_LIKELY(have > n)) {
+    a->ptr += n;
+  } else {
+    a->ptr = UPB_PTRADD(a->ptr, have);
+    a->overflow += (n - have);
+  }
+}
+
+static size_t upb_FieldPath_NullTerminate(upb_PrintfAppender *d, size_t size) {
+  size_t ret = d->ptr - d->buf + d->overflow;
+
+  if (size > 0) {
+    if (d->ptr == d->end) d->ptr--;
+    *d->ptr = '\0';
+  }
+
+  return ret;
+}
+
+static void upb_FieldPath_PutMapKey(upb_PrintfAppender *a, upb_msgval map_key,
+                                    const upb_fielddef *key_f) {
+  switch (upb_fielddef_type(key_f)) {
+    case UPB_TYPE_INT32:
+      upb_FieldPath_Printf(a, "[%" PRId32 "]", map_key.int32_val);
+      break;
+    case UPB_TYPE_INT64:
+      upb_FieldPath_Printf(a, "[%" PRId64 "]", map_key.int64_val);
+      break;
+    case UPB_TYPE_UINT32:
+      upb_FieldPath_Printf(a, "[%" PRIu32 "]", map_key.uint32_val);
+      break;
+    case UPB_TYPE_UINT64:
+      upb_FieldPath_Printf(a, "[%" PRIu64 "]", map_key.uint64_val);
+      break;
+    case UPB_TYPE_BOOL:
+      upb_FieldPath_Printf(a, "[%s]", map_key.bool_val ? "true" : "false");
+      break;
+    case UPB_TYPE_STRING:
+      upb_FieldPath_Printf(a, "[\"");
+      for (size_t i = 0; i < map_key.str_val.size; i++) {
+        char ch = map_key.str_val.data[i];
+        if (ch == '"') {
+          upb_FieldPath_Printf(a, "\\\"");
+        } else {
+          upb_FieldPath_Printf(a, "%c", ch);
+        }
+      }
+      upb_FieldPath_Printf(a, "\"]");
+      break;
+    default:
+      UPB_UNREACHABLE();  // Other types can't be map keys.
+  }
+}
+
+size_t upb_FieldPath_ToText(upb_FieldPathEntry** path, char* buf, size_t size) {
+  upb_FieldPathEntry *ptr = *path;
+  upb_PrintfAppender appender;
+  appender.buf = buf;
+  appender.ptr = buf;
+  appender.end = UPB_PTRADD(buf, size);
+  appender.overflow = 0;
+  bool first = true;
+
+  while (ptr->field) {
+    const upb_fielddef *f = ptr->field;
+
+    upb_FieldPath_Printf(&appender, first ? "%s" : ".%s", upb_fielddef_name(f));
```

Comment:
complex expressions in printf tends to defeat the checker.  this one is simple enough that you are fine, but you should like avoid them where convenient

Reply:
Acknowledged.